docs: rfc-0313 describe validator node registration process (#49)
tari-project/rfcs: src/RFC-0313_VNRegistration.md

|:------------------------|:------------|:------------|
| `VNConfirmationPeriod`  | 1000 Blocks |             |

TODO: explain how `VNConfirmationPeriod` could be changed:
- Changes to `VNConfirmationPeriod` should never be needed, but
- should be based on epoch 
- if more than current, we pause the scanning until the new value is reached
- if less than current we may skip epochs 

## Validator Registration

A validator node operator wishing to participate in [DAN] consensus MUST generate a [Ristretto] keypair `<VN_Public_Key, VN_Secret_Key>` that serves as a stable [DAN] identity and a signing key for L2 consensus messages.
The `VN_Public_Key` MUST be registered on the base layer by submitting a `ValidatorNodeRegistration` [UTXO] which allows the validator to participate in [DAN] consensus for `VNRegistrationValidityPeriod`.

The published `ValidatorNodeRegistration` [UTXO] has these requirements:
1. MUST contain a valid [Schnorr signature] that proves knowledge of the `VN_Secret_Key` 
    * The signature challenge is defined as $e = H(R \mathbin\Vert P \mathbin\Vert m)$
    * $R$ is a public nonce, $P$ is the public `VN_Public_Key` 

>     * $m$ is TODO (previous shard key?, nothing?)

Reviewer: I suggest using the commitment of the UTXO.

point practically impossible. Validators simply ignore [base layer] [reorg]s with a depth _less than_ `VNConfirmationPeriod` deep as the 
data they have extracted is still valid. This means that the _point of finality_ is not always `VNConfirmationPeriod` blocks away from the
tip and is non-decreasing/monotonic which effectively negates [reorg] "noise" from the chain tip. 

However, this does not address [base layer] latency delays where a single huge block or multi-block reorgs may take seconds to be received 
and processed by all [base node]s. Moreover, the validators may poll the base layer only every few seconds, further increasing the latency for 
validators to become aware of the state. This means a single strict validator epoch change-over point will almost always cause liveness failures at 
and after the epoch transition.

To address this, we define a `VNEpochGracePeriod` where both the previous and current epoch are accepted. This value, in blocks, must allow enough
time for all validators to become aware of the [base layer] state for the epoch.

To illustrate, consider the following view of a [base layer] chain. We mark 3 views of the chain.

> ```text

Reviewer 2: We need to take a look here wat happens with leader changes, on point b, we need to make a rule surrounding what leader to listen to, we might have two leaders in E11 and E12, both honest. 

which 2 do the nodes listen to? Their local epoch?

Reviewer: As long as they only vote for one message out of the two leaders, it is fine

                                      (a) (b) (c)                
                                       |   |   |                                         {noisy}
------ | --x------------ | --x------------ | --x------------ | --x------------ | --- .... ------> tip
  |  ϵ10       |   |    ϵ11       |      ϵ12       |       ϵ13               ϵ14
 V_1          V_2 V_3   V_4      V_5              V_6                      
Key:
x - Epoch transition point
ϵn - Epoch n
V_n - Validator registrations 
```

Point (a)
- the validator node set is incomplete for epoch 12.
- the active epoch is 11.
- the validator MUST reject instructions for epoch 12.

Point (b) - the start of epoch 12: 
- the validator nodeset is final for epoch 12.

> - the active epoch remains at 11.

Reviewer: Maybe: VN leaders SHOULD propose using epoch 12, but MAY choose to propose using epoch 11

> - the active epoch remains at 11.

Author: If nodes are behind they will reject 12, 12 kicks in after the transition period.

and spans all blocks from the start of the epoch up to but excluding the start of the next epoch.

*Base layer consensus constants* (all values TBD):

| Name                           | Value              | Description                                                                                |
|:-------------------------------|:-------------------|:-------------------------------------------------------------------------------------------|
| `EpochLength`                  | 60 Blocks (~2h)    | The number of blocks in an epoch                                                           |
| `VNRegistrationValidityPeriod` | 20 Epochs (~40hrs) | The number of epochs that a validator node registration is valid                           |
| `VNRegDepositAmount`           | TBD Tari           | The minimum amount that must be spent to create a valid `ValidatorNodeRegistration` [UTXO] |
| `VNRegLockHeight`              | 10 Epochs          | The lock height that must be set on every `ValidatorNodeRegistration` [UTXO]               |
| `VNShardShuffleInterval`       | 100 Epochs         | The interval that a validator node shard key is shuffled                                   |

> | `VNRegInstateDelay`            | 2 Epochs           | The number of epochs to delay for before instating a new VN                                |

Reviewer: Suggestion: Activation instead of instate
```suggestion
| `VNRegActivationDelay`            | 2 Epochs           | The number of epochs to delay for before activating a new VN                                |
```

    * The signature challenge is defined as $e = H(R \mathbin\Vert P \mathbin\Vert m)$
    * $R$ is a public nonce, $P$ is the public `VN_Public_Key` 
    * $m$ is TODO (previous shard key?, nothing?)
1. the [UTXO]'s `minimum_value` must be at least `VNRegDepositAmount` to mitigate spam/[Sybil] attacks,
1. the [UTXO] lock-height must be set to `VNRegLockHeight`.
1. A script which burns the validator node registration funds if the validator does not reclaim it for a long period after the lock height expires.

```text
OP_PUSH_INT(N) OP_COMPARE_HEIGHT OP_LTE_ZERO OP_IF_THEN 
    OP_PUSH_PUBKEY(VN_Public_Key)
OP_ELSE
    OP_RETURN
OP_END_IF
```

By submitting this [UTXO] the validator node operator is committing to providing a highly-available node for `VNRegistrationValidityPeriod` epochs.

A validator node operator MAY re-register their `VN_Public_Key` before the `VNRegistrationValidityPeriod` epoch is reached, OPTIONALLY 
spending the previous `ValidatorNodeRegistration` [UTXO]. If the previous `ValidatorNodeRegistration` [UTXO] has not expired and
a new `ValidatorNodeRegistration` [UTXO] is submitted, the new `ValidatorNodeRegistration` [UTXO] supersedes the previous one.

> This allows a validator to re-register earlier than the lock-height as long as they are willing to commit 

Reviewer: Is this possibly not the end of the sentence?

time for all validators to become aware of the [base layer] state for the epoch.

To illustrate, consider the following view of a [base layer] chain. We mark 3 views of the chain.

```text
                                      (a) (b) (c)  (d)              
                                       |   |   |    |                                    {noisy}
------ | --x----*------- | --x----*------- | --x----*------- | --x----*------- | --- .... ------> tip
  |  ϵ10       |   |    ϵ11       |      ϵ12       |       ϵ13               ϵ14
 V_1          V_2 V_3   V_4      V_5              V_6                      
Key:
* - Epoch mid-point
x - Epoch transition point
ϵn - Epoch n
V_n - Validator registrations 
```

Point (a)
- the validator node set is incomplete for epoch 12.
- the active epoch is 11.
- the validator MUST reject instructions for epoch 12.

Point (b) - the start of epoch 12: 
- the validator nodeset is final for epoch 12.
- the active epoch remains at 11.
- the validator MUST accept instructions from epoch 11 and 12.

Point (c) - the transition point for epoch 12:
- the active epoch is now 12

> - the validator MUST accept instructions from 11.

Reviewer: ```suggestion
- the validator MUST accept instructions from epoch 11 and 12.
```

Reviewer: it is not clear what should happen between C and D. Is it the rules from C or the rules from D?

Author: From point (c) after the transition point, only epoch 12 is allowed - so we allow a few blocks after the epoch where nodes can be behind. I should have removed point (d) actually, there is no change. The mid point isnt really used - you can set the transition point to be half of the epoch which is the same thing.

- it is assumed at this point that almost all ($3f + 1$) nodes are now on epoch 12

Point (d) - the epoch mid-point:
- the active epoch is 12.
- the validator MUST reject instructions from epoch 11.
- the validator MUST accept instructions from epoch 12.

#### Validator Node Set Definition

> The function $\text{get_vn_set}(\epsilon_n) \rightarrow \vec{S}$ that returns an ordered vector $\vec{S}$ of `VN_Shard_Key`s that are registered for the epoch $\epsilon_n$.

Reviewer: Not sure if this will show up in the mdbook version, but Github shows an error rendering this:
![image](https://user-images.githubusercontent.com/4200336/201323273-fd33ef7f-5643-402d-8b1f-1099b004f078.png)

Author: oh that's annoying, yeah it renders fine in mdbook

}
```

Only a random fraction of validators will be re-assigned shard keys per epoch and that fraction will not be shuffled again
for `VNShardShuffleInterval` epochs. Although the exact number of validators that shuffle per epoch varies, on average 
the `VNShardShuffleInterval` should aim to shuffle around 5% of the network at every epoch. This is to ensure that the

> number of shuffling validators is far from $\frac{1}{3}$ of the validator set, as this could break liveness and safety 

Reviewer: ```suggestion
number of shuffling validators is much less than $\frac{1}{3}$ of the validator set, as this could break liveness and safety 
```
